RFC.md

```diff
@@ -0,0 +1,253 @@
+
+        # shape [num_experts,]: how many tokens for each expert
+        num_local_tokens_per_expert = torch.histc(selected_expert_indices.view(-1), bins=num_experts, min=0, max=num_experts)
+        # shape [bs*slen*experts_per_token,]
+        token_indices_experts_sorted = torch.argsort(selected_experts_indices.view(-1), stable=True)
+        # top_scores shape [bs*slen*experts_per_token,]
+        top_scores = top_scores.view(-1)[token_indices_experts_sorted]
+
+        # token_indices shape [bs*slen*experts_per_token, hidden_dim]
+        token_indices = token_indices_experts_sorted.reshape(-1, 1).expand(-1, hidden_dim)
+        # routed_input shape [bs*slen*experts_per_token, hidden_dim]
+        routed_input = torch.gather(x, dim=0, index=token_indices)
+        routed_input = routed_input * top_scores
+
+        return routed_input, token_indices, num_local_tokens_per_expert
+ ```
+
+However, token choice routing has several pitfalls according to the expert choice [paper](https://arxiv.org/pdf/2002.05202).
+1. Poor load balance. Experts can become under or over-specialized. Load imbalance can hurt step latency / inference time.
+2. Experts under specialization. Ideally the gating network will learn token-to-expert affinity such that similar or relevant tokens are routed to the same expert. However, a sub-optimal strategy can produce redundant experts and/or experts that are not sufficiently specialized.
+3. Same compute for each token. Token choice will allocate a fixed number of experts to each token regardless of the importance of different tokens. Ideally an MOE model should flexibly allocate compute resources based on the complexity of the input.
+
+Compared to **token choice**, **expert choice** topK routing lets experts select its top-k tokens. The ExpertChoiceTopKRouter class routes input tokens to different experts based on the router scores.
+
+**Here's the proposed Expert Choice Routing design in torchtune:**
+```python
+class ExpertChoiceTopKRouter(nn.Module):
+    def __init__(self, hidden_dim, num_experts):
+        self.gate = nn.Linear(hidden_dim, num_experts)
+        self.tokens_per_expert = tokens_per_expert
+
+    def forward(self, x, use_sigmoid=False):
+        '''
+        input:
+            x: shape [bs*slen, hidden_dim]
+        outputs:
+            routed_input: selected tokens
+                shape [num_experts*tokens_per_expert, hidden_dim]
+            token_indices: selected token indices
+            num_local_tokens_per_expert: None
+        '''
+        # scores shape [num_experts, bs*slen]
+        scores = self.gate(x).transpose(0,1)
+        if use_sigmoid:
+            scores = torch.sigmoid(scores.to(sigmoid_dtype)).to(x.dtype)
+        else:
+            scores = F.softmax(scores.to(softmax_dtype), dim=0).to(x.dtype)
+        # router scores/indices shape [num_experts, tokens_per_expert]
+        top_scores, selected_token_indices = torch.topk(scores, k=self.tokens_per_expert, dim=1)
```

Comment:
Discussed offline a bit but we should make sure there isn't any information leakage happening here for decoder models. Specifically, this line seems a bit fishy to me. It seems like each expert is looking across all tokens and choosing its top tokens_per_expert based on the entire sequence, right?